### PR TITLE
refactor: split queue process out

### PR DESCRIPTION
**Pull request by shreddedbacon**
# Checklist

- [ ] Affected Issues have been mentioned in the Closing issues section
- [ ] Documentation has been written/updated
- [ ] PR title is ready for changelog and subsystem label(s) applied

Refactoring build queue process to reduce the amount of times that events trigger actions when QoS is enabled.

There were points in the event cycle where when a build completes, it would pick an item from the queue, but if there was also a pending build in the namespace that triggered the event, it would start a build in that namespace too. This allowed clusters to exceed the QoS limit.

This also introduces some additional changes to the way the queue is handled with the sorting function working as it should now by sorting by priority and creation time, and also allows events to trigger period updates of the pending items in a blocking operation

### CODE REVIEW

File: controllers/v1beta1/build_helpers.go [hunk 1]
```diff
@@ -731,7 +731,7 @@ func (r *LagoonBuildReconciler) processBuild(ctx context.Context, opLog logr.Log
 			VolumeSource: corev1.VolumeSource{
 				Secret: &corev1.SecretVolumeSource{
 					SecretName:  "lagoon-sshkey",
-					DefaultMode: helpers.IntPtr(420),
+					DefaultMode: helpers.Int32Ptr(420),
```
Comment (bomoko) [inline controllers/v1beta1/build_helpers.go:734]: NOTE for self: okay - so practically this doesn't change here, we're just renaming a function to be more precise.

File: controllers/v1beta1/build_qoshandler.go [hunk 2]
```diff
@@ -52,18 +51,42 @@ func (r *LagoonBuildReconciler) whichBuildNext(ctx context.Context, opLog logr.L
...
+	}
+	return nil
+}
+
+var runningProcessQueue bool
+
+// this is a processor for any builds that are currently `queued` status. all normal build activity will still be performed
+// this just allows the controller to update any builds that are in the queue periodically
+// if this ran on every single event, it would flood the queue with messages, so it is restricted using `runningProcessQueue` global
+// to only run the process at any one time til it is complete
```
Comment (bomoko) [inline controllers/v1beta1/build_qoshandler.go:76]: might be useful here to give a rundown of the arguments - specifically "buildsToStart" and "limitHit"

File: controllers/v1beta1/build_qoshandler.go [hunk 3]
```diff
@@ -52,18 +51,42 @@ func (r *LagoonBuildReconciler) whichBuildNext(ctx context.Context, opLog logr.L
...
+func (r *LagoonBuildReconciler) processQueue(ctx context.Context, opLog logr.Logger, buildsToStart int, limitHit bool) error {
+	// this should only ever be able to run one instance of at a time within a single controller
+	// this is because this process is quite heavy when it goes to submit the queue messages to the api
+	// the downside of this is that there can be delays with the messages it sends to the actual
+	// status of the builds, but build complete/fail/cancel will always win out on the lagoon-core side
+	// so this isn't that much of an issue if there are some delays in the messages
+	opLog = opLog.WithName("QueueProcessor")
```
Comment (bomoko) [inline controllers/v1beta1/build_qoshandler.go:83]: as an aside, not part of the review, but I wonder if it wouldn't be cool to have the logger itself contain the logic for whether debug is enabled or not, so you're just calling a function

just a thought really

```
	enableDebug := true

	logger := func(l string) {
		if enableDebug {
			fmt.Println(l)
		}
	}

	logger("hello world")

	enableDebug = false

	logger("shouldn't show up")
```

Comment (bomoko) [reply controllers/v1beta1/build_qoshandler.go:83]: to be clear, this is absolutely not part of the review, just me thinking out loud :D

I know we're calling subfunctions of the logger class, that complicates it. It's likely just simpler to do the damned `if`

File: controllers/v1beta1/build_qoshandler.go [hunk 4]
```diff
@@ -52,18 +51,42 @@ func (r *LagoonBuildReconciler) whichBuildNext(ctx context.Context, opLog logr.L
...
 	if buildsToStart > 0 {
 		opLog.Info(fmt.Sprintf("Currently %v running builds, room for %v builds to be started", len(runningBuilds.Items), buildsToStart))
 		// if there are any free slots to start a build, do that here
-		listOption = (&client.ListOptions{}).ApplyOptions([]client.ListOption{
+		go r.processQueue(ctx, opLog, buildsToStart, false)
+	}
+	return nil
+}
+
+var runningProcessQueue bool
```
Comment (bomoko) [inline controllers/v1beta1/build_qoshandler.go:71]: This seems a prime candidate for replacing with a mutex and a defer to clear.

I wonder, does this ever potentially cause issues as it's currently implemented?
I see that we set this to true at the beginning of the process, but then in some cases the process returns without setting it to false again - how does this recover? (does it matter?)

**Review by bomoko — CHANGES_REQUESTED** [review]
Looking good, don't necessarily have changes - but questions, specifically about the locking mechanism.

File: controllers/v1beta1/build_qoshandler.go (hunk 4 above)
Comment (shreddedbacon) [reply controllers/v1beta1/build_qoshandler.go:71]: Yeah, might have missed some resets in there. Will check.

I'm happy to look at replacing it with a mutex with defer, but previous attempts lead the process to just be delayed execution, not prevention of execution

File: controllers/v1beta1/build_qoshandler.go (hunk 3 above)
Comment (shreddedbacon) [reply controllers/v1beta1/build_qoshandler.go:83]: I didn't want to complicate things with that, so the `if` is just simpler

File: controllers/v1beta1/build_helpers.go (hunk 1 above)
Comment (shreddedbacon) [reply controllers/v1beta1/build_helpers.go:734]: Yeah, just a rename because I had to re-implement `IntPtr` to do `int` and it was set for `int32`

**Review by bomoko — APPROVED** [review]
Looks good!

**Pull request merged**